---
title: "Filter evaluation"
author: "Harly Jane Durbin"
output:
  html_document:
    toc: true
    df_print: paged
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(dplyr)
library(tidyr)
library(readr)
library(magrittr)
library(ggplot2)

source(here::here("source_functions/filter_eval_functions.R"))
```

# Notes & questions

* From GATK [`VariantsToTable` docs](https://gatk.broadinstitute.org/hc/en-us/articles/360036896892-VariantsToTable#--genotype-fields):
    + TRANSITION (1 for a bi-allelic transition (SNP), 0 for bi-allelic transversion (SNP), -1 for INDELs and multi-allelics)
    + NO-CALL (count of no-call genotypes)
    + VAR (count of non-reference genotypes)
* Need to plot
    + Depth of coverage per individual
    + Mean coverage per site (Yassine says limit to 8)
    + GQ per site (Yassine says limit to 20)
* Re: Phred scale: "For many purposes, a Phred Score of 20 or above is acceptable, because this means that whatever it qualifies is 99% accurate, with a 1% chance of error."
* 2,764 samples pre-filtering, 2,362 samples post-filtering

`shuf -n N input > output`

# Setup 

```{r}
sample_metadata <- read_csv(here::here("data/derived_data/metadata/bovine_demo.sample_metadata.csv"))
```


```{r}
eval_28 <- 
  c("data/derived_data/joint_genotyping/filter_eval/filter_eval.with_indels.table", "data/derived_data/joint_genotyping/filter_eval/filter_eval.post.table") %>% 
  purrr::set_names(c("pre", "post")) %>% 
  purrr::map_dfr(~ read_table2(here::here(.x)),
             .id = "stage") %>% 
  janitor::clean_names() %>% 
  left_join(
    read_table2(here::here("data/derived_data/joint_genotyping/filter_eval/filter_eval.with_indels.28.ldepth.mean")) %>% 
      select(-CHROM) %>% 
      janitor::clean_names())
```

```{r}
nocalls <-
  eval_28 %>% 
  filter(type == "SNP") %>% 
  select(stage, pos, no_call) 
```

```{r}
gq <-
  read_table2(here::here("data/derived_data/joint_genotyping/filter_eval/filter_eval.28_random.GQ.txt"), na = ".") %>% 
  tidyr::pivot_longer(cols = all_of(3:2766), names_to = "international_id", values_to = "gq") %>% 
  janitor::clean_names() %>% 
  filter(international_id %in% sample_metadata$international_id) %>% 
  select(-chrom)
```


# How many SNPs within 5 bp of an indel?

```{r}
indel_range <- 
  eval_28 %>% 
  filter(type == "INDEL") %>% 
  select(pos) %>% 
  mutate(list = purrr::map(.x = pos, ~ seq(from = .x-5, to = .x+5))) %>% 
  unnest(list) %>% 
  pull(list)
```

```{r, echo=FALSE}
in_range_length <-
  eval_28 %>% 
  filter(stage == "pre") %>% 
  filter(type == "SNP") %>% 
  filter(pos %in% indel_range) %>% 
  pull(pos) %>% 
  length(.) 

snp_length <-
  eval_28 %>%
  filter(stage == "pre") %>% 
  filter(type == "SNP") %>%
  pull(pos) %>%
  length(.)
  
```

`r in_range_length %>% scales::comma(.)` of `r snp_length %>% scales::comma(.)` total SNPs are within 5 bp of an indel (`r scales::percent(in_range_length/snp_length)`).

# Filter values prior to QC

## Quality by depth

* "[QUAL](https://gatkforums.broadinstitute.org/gatk/discussion/1268/what-is-a-vcf-and-how-should-i-interpret-it): The Phred-scaled probability that a REF/ALT polymorphism exists at this site given sequencing data. Because the Phred scale is -10 x log(1-p), a value of 10 indicates a 1 in 10 chance of error, while a 100 indicates a 1 in 10^10 chance (see the FAQ article for a detailed explanation). These values can grow very large when a large amount of data is used for variant calling, **so QUAL is not often a very useful property for evaluating the quality of a variant call.** 
Not to be confused with the sample-level annotation GQ; see this FAQ article for an explanation of the differences in what they mean and how they should be used."
* "[QUAL](https://gatk.broadinstitute.org/hc/en-us/articles/360035531392-Difference-between-QUAL-and-GQ-annotations-in-germline-variant-calling) tells you how confident we are that there is some kind of variation at a given site. The variation may be present in one or more samples."

```{r}
eval_28 %>% 
  filter(type == "SNP") %>% 
  plot_filter_density(var = qd, plot_title = "Density of QD values, chromosome 28") + 
  geom_vline(xintercept = 10, color = "red")
```

## Fisher strand

* "[Strand bias](https://gatk.broadinstitute.org/hc/en-us/articles/360036361992-FisherStrand) is a type of sequencing bias in which one DNA strand is favored over the other, which can result in incorrect evaluation of the amount of evidence observed for one allele vs. the other. The output is a Phred-scaled p-value. The higher the output value, the more likely there is to be bias. More bias is indicative of false positive calls. FisherStrand is best suited for low coverage situations. For testing strand bias in higher coverage situations, see the StrandOddsRatio annotation."
* Have used FS > 60 in the past

```{r}
eval_28 %>% 
  filter(type == "SNP") %>% 
  plot_filter_density(var = fs, plot_title = "Density of FS values, chromosome 28") + 
  geom_vline(xintercept = 60, color = "red")
```

The GATK team [reccomends](http://mbontrager.org/blog/2016/08/17/Variant-Exploration) log(10) scaling the x-axis when plotting FS:

```{r}
eval_28 %>% 
  filter(type == "SNP") %>% 
  plot_filter_density(var = fs, plot_title = "Density of FS values, chromosome 28") +
  scale_x_continuous(trans = 'log10') +
  labs(x = "log10(FS)")
```

## RMS mapping quality

* "[Root mean square mapping quality](https://gatk.broadinstitute.org/hc/en-us/articles/360037591751-RMSMappingQuality) provides an estimation of the overall mapping quality of reads supporting a variant call, averaged over all samples in a cohort."
* Have used MQ < 15 in the past

```{r}
eval_28 %>% 
  filter(type == "SNP") %>% 
  plot_filter_density(var = mq, plot_title = "Density of MQ values, chromosome 28") +
  geom_vline(xintercept = 15, color = "red")
```

## Mapping quality rank sum test

* Have used MQRankSum < -12.5 in the past

```{r}
eval_28 %>% 
  filter(type == "SNP") %>% 
  plot_filter_density(var = mq_rank_sum, plot_title = "Density of MQRankSum values, chromosome 28") +
  geom_vline(xintercept = -12.5, color = "red")
```

## Read position rank sum

* Have used ReadPosRankSum < -8.0 in the past

```{r}
eval_28 %>% 
  filter(type == "SNP") %>% 
  plot_filter_density(var = read_pos_rank_sum, "Density of ReadPosRankSum values, chromosome 28") +
  geom_vline(xintercept = -8)
```

## Mean depth per site averaged across all individuals

```{r}
eval_28 %>% 
  filter(type == "SNP") %>% 
  plot_filter_density(var = mean_depth, "Mean depth per site, chromosome 28") +
  geom_vline(xintercept = 8, color = "red")
```

# Summary

```{r}

eval_28 %>%
  filter(type == "SNP") %>%
  summarize_dropped(var = mean_depth, cutoff = 8) %>%
  bind_rows(
    eval_28 %>%
      filter(type == "SNP") %>%
      summarize_dropped(var = qd, cutoff = 10)) %>% 
  bind_rows(
    eval_28 %>%
      filter(type == "SNP") %>%
      summarize_dropped(var = mq, cutoff = 15)) %>% 
  bind_rows(
    eval_28 %>%
      filter(type == "SNP") %>%
      summarize_dropped(var = mq_rank_sum, cutoff = -12.5)) %>% 
  bind_rows(
    eval_28 %>%
      filter(type == "SNP") %>%
      summarize_dropped(var = read_pos_rank_sum, cutoff = -8)) 
```

## What percentage of SNPs actually were dropped?

```{r}
nocalls %>% 
  tidyr::pivot_wider(values_from = no_call, names_from = stage) %>% 
  summarise(`Percent dropped` = sum(is.na(post))/sum(!is.na(pre)))
```

# No-calls

* 7/14/2020: Removing SNPs with 10% or greater missingness removed nearly all SNPs. What went wrong?

## How many SNPs were actually set to missing by genotype-level filtering?

```{r}
nocalls %>% 
  filter(!is.na(no_call)) %>% 
  mutate(percent_missing = 
           case_when(
             stage == "post" ~ no_call/2362,
             stage == "pre" ~ no_call/2764
             )) %>% 
  group_by(stage) %>% 
  summarise(`Min missing` = min(percent_missing),
            `Max missing` = max(percent_missing),
            `Avg missing` = mean(percent_missing))
```

```{r}
nocalls %>% 
  mutate(percent_missing = 
           case_when(
             stage == "post" ~ no_call/2362,
             stage == "pre" ~ no_call/2764
             )) %>% 
  ggplot(aes(x = percent_missing)) +
  geom_histogram() +
  facet_wrap(~ stage, nrow = 2, scales = "free_y")
```

## Re-evaluate starting genotype quality using 5,000 randomly selected chromosome 28 SNPs

* "[GQ](https://gatk.broadinstitute.org/hc/en-us/articles/360035531692) represents the Phred-scaled confidence that the genotype assignment (GT) is correct, derived from the genotype PLs. *Specifically, the GQ is the difference between the PL of the second most likely genotype, and the PL of the most likely genotype.* Basically the GQ gives you the difference between the likelihoods of the two most likely genotypes. If it is low, you can tell there is not much confidence in the genotype, i.e. there was not enough evidence to confidently choose one genotype over another."

### Summary across all sites and all individuals

```{r}
gq %>% 
  mutate(gq = if_else(20 > gq, NA_integer_, as.integer(gq))) %>% 
  summarise(`% genotypes with GQ < 20` = sum(is.na(gq))/(sum(is.na(gq))+sum(!is.na(gq))))
```

```{r}
gq %>% 
  mutate(stage = "pre") %>% 
  plot_filter_density(gq) +
  geom_vline(xintercept = 20, color = "red") +
  labs(title = "Density of GQ values, all genotypes at all sites")
```

### Summary averaged by site

```{r}
gq %>% 
  group_by(pos) %>% 
  summarise(gq = mean(gq, na.rm = TRUE)) %>% 
  mutate(stage = "pre") %>% 
  plot_filter_density(gq) +
  geom_vline(xintercept = 20, color = "red") +
  labs(title = "Density of GQ values, averaged by site")
```

### How many sites where > 10% of samples with GQ < 20?

```{r}
gqfail <-
  gq %>% 
  mutate(dropped = if_else(20 > gq, "TRUE", NA_character_)) %>% 
  group_by(pos) %>% 
  summarise(percent_fail = sum(!is.na(dropped))/(sum(!is.na(dropped))+sum(is.na(dropped)))) %>% 
  filter(percent_fail > 0.1) %>% 
  ungroup() %>% 
  pull(pos) %>% 
  length(.)
```

Of 5,000 randomly pulled sites on chr 28, `r gqfail %>% scales::comma(.)` where > 10% of the samples would fail/have GQ < 20

```{r}
gq %>% 
  mutate(dropped = if_else(20 > gq, "TRUE", NA_character_)) %>% 
  group_by(pos) %>% 
  summarise(percent_fail = sum(!is.na(dropped))/(sum(!is.na(dropped))+sum(is.na(dropped)))) %>%
  ungroup() %>% 
  summarise(`Min % fail` = min(percent_fail), 
            `Max % fail` = max(percent_fail),
            `Mean % fail` = mean(percent_fail))
```

```{r}
gq %>% 
  mutate(dropped = if_else(20 > gq, "TRUE", NA_character_)) %>% 
  group_by(pos) %>% 
  summarise(percent_fail = sum(!is.na(dropped))/(sum(!is.na(dropped))+sum(is.na(dropped)))) %>%
  ungroup() %>% 
  ggplot(aes(x = percent_fail)) +
  geom_histogram()
```


# Commentary
