---
title: "Joint genotyping"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: show
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(readr)
library(dplyr)
library(ggplot2)
library(purrr)
library(tidyr)
library(tidylog)
```

# Notes & questions

* [CollectVariantCallingMetrics](https://broadinstitute.github.io/picard/picard-metric-definitions.html#CollectVariantCallingMetrics.VariantCallingDetailMetrics)
    + HET_HOMVAR_RATIO: (count of hets)/(count of homozygous non-ref) for this sample
    + SNP_REFERENCE_BIAS: The rate at which reference bases are observed at ref/alt heterozygous SNP sites.

# Setup

```{r}
sample_metadata <- read_csv(here::here("data/derived_data/metadata/bovine_demo.sample_metadata.csv"))
```

```{r}
var_detail <-
  c(1:29, "X", "Y") %>% 
  purrr::set_names() %>% 
  purrr::map_dfr(~ read_table2(here::here(glue::glue("data/derived_data/joint_genotyping/collect_metrics/collect_metrics.{.x}.variant_calling_detail_metrics")), skip = 6, col_types = cols(.default = "c")), .id = "chr") %>% 
  janitor::clean_names() %>% 
  rename(international_id = sample_alias) %>% 
  # Removed indels, removed multi-allelics, physically removed SNPs that failed filtration
  select(-contains("indels"), -contains("ins_del"), -contains("multiallelic"), -filtered_snps) %>% 
  right_join(sample_metadata %>% 
              select(international_id, lab_id, population, species, avg_coverage)) %>% 
  mutate_at(.vars = vars(het_homvar_ratio:num_singletons), ~ as.numeric(.)) %>% 
  select(chr, population, avg_coverage, everything())
```

```{r}
var_detail_wg <-
  var_detail %>%
  group_by(international_id, lab_id, avg_coverage, population, species) %>%
  summarise_at(vars(
    het_homvar_ratio,
    pct_dbsnp,
    dbsnp_titv,
    novel_titv,
    snp_reference_bias
  ),
  mean, na.rm = TRUE) %>%
  ungroup() %>%
  full_join(
    var_detail %>%
      group_by(international_id, lab_id, avg_coverage, population, species) %>%
      summarise_at(vars(
        total_snps,
        num_in_db_snp,
        novel_snps,
        num_singletons
        ),
        sum) %>%
      ungroup()
    )

```

```{r}
var_sum <-
  c(1:29, "X", "Y") %>% 
  purrr::set_names() %>% 
  purrr::map_dfr(~ read_table2(here::here(glue::glue("data/derived_data/joint_genotyping/collect_metrics/collect_metrics.{.x}.variant_calling_summary_metrics")), skip = 6), .id = "chr") %>% 
  janitor::clean_names() %>% 
  select(-contains("indels"), -contains("ins_del"), -contains("multiallelic"), -filtered_snps) # %>% 
 # mutate_at(c("total_snps", "num_in_db_snp", "novel_snps", "num_singletons"), ~ scales::comma(.)) 
```

# Variant calling summary metrics

```{r}
var_sum %>% 
  summarise(sum(total_snps))
```

# Variant calling detail metrics

```{r}
var_detail_wg %>% 
  arrange(desc(novel_titv))

```

```{r}
var_detail_wg %>% 
  group_by(population) %>% 
  filter(n() > 50) %>%
  ungroup() %>% 
  mutate(filter = if_else(10 > avg_coverage, "FAIL", "PASS")) %>% 
  group_by(population, filter) %>% 
  tally() %>% 
  ungroup() %>% 
  tidyr::pivot_wider(names_from = filter, values_from = n)
```


```{r}
sample_metadata %>% 
  filter(str_detect(population, "yak ")) %>% 
  group_by(population) %>% 
  tally()

sample_metadata %>% 
  filter(population == "yak datong")
```

