---
title: "Memory tracking"
author: "Harly Durbin"
output:
  html_document:
    toc: true
    toc_depth: 2
    df_print: paged
    code_folding: show
---

```{r setup}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning = FALSE)
library(readr)
library(dplyr)
library(ggplot2)
library(purrr)
library(tidyr)
library(lubridate)

source(here::here("source_functions/psrecord_summary.R"))
```

# `ancient_preprocess`

## `trimBam`

```{r}
usage_summary(process = "ancient_preprocess", keyword = "trimBam")
```

```{r}
usage_facets(process = "ancient_preprocess", keyword = "trimBam")
```

## `index_bam`

```{r}
usage_summary(process = "ancient_preprocess", keyword = "index_bam")
```

```{r}
usage_facets(process = "ancient_preprocess", keyword = "index_bam")
```

## `target_creator`

```{r}
usage_summary(process = "ancient_preprocess", keyword = "target_creator")
```

```{r}
usage_facets(process = "ancient_preprocess", keyword = "target_creator")
```

## `indel_realigner`

```{r}
usage_summary(process = "ancient_preprocess", keyword = "indel_realigner")
```

```{r}
usage_facets(process = "ancient_preprocess", keyword = "indel_realigner")
```

## `merge_realigned`

```{r}
usage_summary(process = "ancient_preprocess", keyword = "merge_realigned")
```

```{r}
usage_facets(process = "ancient_preprocess", keyword = "merge_realigned")
```

# `joint_genotyping`

* ~ 2,900 multi-breed and multi-species *Bos* samples

## `combine_gvcfs`

* Combining 33 cohort-GVCFs of ~ 90 samples each into chromosome-level GVCFs using `GATK CombineGVCFs`
* For each chromosome, provided 4 threads

```{r}
usage_summary(process = "joint_genotyping", keyword = "combine_gvcfs") 
```

## `genotype_gvcfs`

* Chromosome-by-chromosome joint genotyping with `GATK GenotypeGVCFs`
* For each chromosome, provided 28 threads

```{r}
usage_summary(process = "joint_genotyping", keyword = "genotype_gvcfs") 
```

```{r}
usage_facets(process = "joint_genotyping", keyword = "genotype_gvcfs")
```

## `select_variants`

* Removing indels & multiallelic SNPs using `GATK SelectVariants`
* For each chromosome, provided 8 threads

```{r}
usage_summary(process = "joint_genotyping", keyword = "select_variants") 
```

```{r}
usage_facets(process = "joint_genotyping", keyword = "select_variants")
```

## `variant_filtration`

* Site-level and genotype-level filtering using `GATK VariantFiltration`
* Did not multi-thread beyond java GC threads

```{r}
usage_summary(process = "joint_genotyping", keyword = "variant_filtration") 
```

```{r}
usage_facets(process = "joint_genotyping", keyword = "variant_filtration")
```

## `remove_failed`

* Removing variants that didn't pass filtering using `GATK SelectVariants`
* For each chromosome, provided 8 threads

```{r}
usage_summary(process = "joint_genotyping", keyword = "remove_failed") 
```

```{r}
usage_facets(process = "joint_genotyping", keyword = "remove_failed")
```

# `post_process`

## `remove_samples`

* Remove duplicate and low quality samples using `GATK SelectVariants`

```{r}
usage_summary(process = "joint_genotyping", keyword = "remove_samples") 
```

```{r}
usage_facets(process = "joint_genotyping", keyword = "remove_samples")
```

## `collect_metrics`

* Run `Picard CollectVariantCallingMetrics` on chromosome files

```{r}
usage_summary(process = "joint_genotyping", keyword = "collect_metrics") 
```

```{r}
usage_facets(process = "joint_genotyping", keyword = "collect_metrics")
```

## `validate_variants`

* Run `GATK ValidateVariants` on chromosome files

```{r}
usage_summary(process = "joint_genotyping", keyword = "validate_variants") 
```

```{r}
usage_facets(process = "joint_genotyping", keyword = "validate_variants")
```

# `sex_imputation`

## `guess_ploidy`

* Use `bcftools +guess-ploidy` plugin to guess sexes of samples

```{r,eval=FALSE}
usage_summary(process = "joint_genotyping", keyword = "guess_ploidy") 
```

```{r, eval=FALSE}
usage_facets(process = "joint_genotyping", keyword = "guess_ploidy")
```

# `phasing`